### ISSUE TITLE
STPPaymentCardTextField sends keyboardWillHideNotification when begin editing the field.

### ISSUE THREAD

**Issue by rus64** [stripe/stripe-ios#1388]
## Summary
When tapping on `STPPaymentCardTextField` to begin editing the field one would expect `UIResponder.keyboardWillShowNotification` to be sent, but instead the `UIResponder.keyboardWillHideNotification` is sent. This results in great difficulty in managing the View in relation to the keyboard.

## Code to reproduce
```swift
NotificationCenter.default.addObserver(self, selector: #selector(keyboardWillHide(_:)), name: UIResponder.keyboardWillHideNotification, object: nil)
```

## iOS version
<= iOS 12

## Installation method
Carthage

## SDK version
17.0.1

**Comment by davidme-stripe** [comment]
Hey @rus64, I'm seeing this happen when a hardware keyboard is connected (or when "Keyboard" -> "Connect Hardware Keyboard" is enabled in the Simulator menu, which is the default option), but not when it's disabled. Without a hardware keyboard, tapping on the field triggers a `UIResponder.keyboardWillShowNotification`. Are you seeing this without a hardware keyboard connected?

**Comment by rus64** [comment]
Hi @davidme-stripe, thanks for your response. Making sure that setting was disabled did indeed avoid the mentioned issue, however, now when dismissing the keyboard both the `UIResponder.keyboardWillShowNotification` and `UIResponder.keyboardWillShowNotification` are sent, causing glitchy behaviour.

**Comment by rus64** [comment]
On further inspection this is the order of events that I am seeing:

- Begin editing cardTextField (`STPPaymentCardTextField`)
- `keyboardWillShowNotification` is sent
- `keyboardDidShowNotification` is sent
- Tap Cancel - this calls `cardTextField.resignFirstResponder()`
- `keyboardWillHideNotification` is sent
- `keyboardDidHideNotification` is sent
- `cardTextField.isFirstResponder` is still true here and the keyboard is visible
-  After didHide notification, call `viewController.dismiss()`
- `keyboardWillShowNotification` is sent
- `keyboardDidShowNotification` is sent
- `keyboardWillHideNotification` is sent

It seems to me that maybe the keyboard is not properly dismissed after resigning the card text field. The last three notifications then cause weird behaviour in our app.

Tested on an iPhone 6s on iOS 12.2 (real device not sim).

**Comment by rus64** [comment]
A temporary workaround that I've found:

Instead of calling `resignFirstResponder` on the Card Text Field, we can call `endEditing(true)` on the ViewController's View, forcing the view's active subview to resign.

From the `UIKit.UITextField` documentation:
> use to make the view or any subview that is the first responder resign (optionally force)

**Comment by davidme-stripe** [comment]
Thanks for the additional info! I've tried to patch this in the branch `davidme/card-text-field-less-noisy`. Let me know if it fixes the issue for you.

**Comment by rus64** [comment]
@davidme-stripe, thanks for taking a look! Was the branch with this fix merged onto `master`?

**Comment by rus64** [comment]
Unfortunately the patch above does not fix the issue for me. Still seeing unexpected keyboard behaviour when calling `resignFirstResponder()` on the `STPPaymentCardTextField`.

**Comment by davidme-stripe** [comment]
@rus64, I can't seem to reproduce the issue after the patch.

* I tap the card field and receive a keyboardWillShowNotification.
* I tap the expiration date field and receive a second keyboardWillShowNotification. This seems like normal UIKit behavior: The keyboard dimensions may change when moving to a different text field, and the `userInfo` of the notification contains the new dimensions.
* I set a timer to wait 5 seconds before calling `cardTextField.resignFirstResponder()`. This sends a single keyboardWillHideNotification.
* I do the same setup, then dismiss the entire view controller. This sends a single keyboardWillHideNotification.

What behavior are you seeing after the new patch?

**Comment by rus64** [comment]
Ah, I do apologise; looks like I did not properly build the framework from the master branch. I tested again this morning and confirm that the issue is fixed. Thank you for your time and attention on this!

**Comment by davidme-stripe** [comment]
This is fixed in 18.0.0. Thanks for filing!
